### PR TITLE
C/C++ : Useless test

### PR DESCRIPTION
**Pull request by Yonah125**
This query finds any useless test of kind a==8 && a!=7.

This query found an useless test on the Suricata project that can you see in this Pull Request : 

https://github.com/OISF/suricata/pull/6897

### CODE REVIEW

**Comment by catenacyber** [comment]
cc @HauwaOtori : this is a first PR from my students project aka CodeQL versus Suricata
cc @xcorail

PS @Yonah125 do you want to co-author commits with your comrades ?

File: cpp/ql/src/experimental/Best Practices/UselessTest.ql
Comment (MathiasVP) [inline cpp/ql/src/experimental/Best Practices/UselessTest.ql:13]: I'm not quite sure I understand the meaning of this predicate. The name makes it sound like you want to check if two statements are equivalent, but the two parameters are of type `Expr` (and not `Stmt`)?

If what you want is to compare two expressions, I'd consider trying out our global value numbering library (by importing `semmle.code.cpp.valuenumbering.GlobalValueNumbering`). Then you can check if two expressions are guarenteed to evaluate to the same value by check if `globalValueNumber(e1).getAnExpr() = e2` holds. We also have a hash cons library which is a more syntactic version of global value numbering. You can read up on both libraries here: https://codeql.github.com/docs/codeql-language-guides/hash-consing-and-value-numbering/

File: cpp/ql/src/experimental/Best Practices/UselessTest.ql
Comment (MathiasVP) [inline cpp/ql/src/experimental/Best Practices/UselessTest.ql:15]: You shouldn't be using the value produced by `toString` in your query logic as we don't guarantees any backwards compatibility on the strings produced by `toString` (i.e., we might silently change the format of the string in the future and thus break your query logic).

File: cpp/ql/src/experimental/Best Practices/UselessTest.ql
Comment (MathiasVP) [inline cpp/ql/src/experimental/Best Practices/UselessTest.ql:41]: I'd avoid using `getChild` whenever there's a more convenient predicate available. For instance, `eq.getChild(0)` is equivalent to `eq.getLeftOperand()`m and `eq.getChild(1)` is equivalent to `eq.getRightOperand()`. The same holds for the uses of `ne.getChild`.

File: cpp/ql/src/experimental/Best Practices/UselessTest.qhelp
Comment (MathiasVP) [inline cpp/ql/src/experimental/Best Practices/UselessTest.qhelp:7]: A couple of grammar fixes :) Also, our PR checks are complaining because of the non-escaped `&&`, so I've replaced it with `&amp;&amp;`.
```suggestion
<p>Comparison operations like <code>a==8 &amp;&amp; a!=7</code> contain a useless part : the non-equal part. This rule finds tests of this kind within an <code>if</code> or a <code>while</code> statement
```

File: cpp/ql/src/experimental/Best Practices/UselessTest.ql
Comment (MathiasVP) [inline cpp/ql/src/experimental/Best Practices/UselessTest.ql:3]: Code must be enclosed in single quotes. See https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md#query-descriptions-description.

Also, ideally the query description should describe the problem that this query is trying to highlight, and not what the query is doing. So I'd really rephrase this as something like:
```
A boolean condition that is guaranteed to never be evaluated should be deleted.
```

**Review by MathiasVP — COMMENTED** [review]
Thanks for your contribution @Yonah125,

Here are my initial comments. Note that CI is unhappy because the query isn't hasn't been auto-formatted (see the `Formatting` section in https://github.com/github/codeql/blob/main/CONTRIBUTING.md#submitting-a-new-experimental-query).

None of my comments are really blocking this PR: It's mostly to help you learn to be efficient writing CodeQL queries :) If you're happy with the results you're finding with the query I'll merge it as soon as the PR is green.

**Comment by MathiasVP** [comment]
> About the auto-formatting, I used the 'Format Document' command in VSC, is the output bad ?

Excellent. That should all be fine now, then. Let's see what CI thinks about it now 🤞.

File: cpp/ql/src/experimental/Best Practices/UselessTest.ql
Comment (MathiasVP) [inline cpp/ql/src/experimental/Best Practices/UselessTest.ql:43]: The PR check is failing because I missed an error in your query :)

Since this is a `@kind problem` query, your `select` clause must be of the form `select element, "string"` (instead of `select "string", element` like you have currently). Check out https://codeql.github.com/docs/writing-codeql-queries/about-codeql-queries/#select-clause for more info on this. This edit should make CI happy:

```suggestion
select ne, "Useless test"
```

File: cpp/ql/src/experimental/Best Practices/UselessTest.ql
Comment (rdmarsh2) [inline cpp/ql/src/experimental/Best Practices/UselessTest.ql:18]: I think you want `getConverted*()` rather than `getFullyConverted` - if there's an explicit cast of a parenthesized expression, like `(short)(a + b)`, `getFullyConverted()` will be the cast rather than the parenthesis.

**Comment by catenacyber** [comment]
CI looks green now 👏

**Comment by MathiasVP** [comment]
> CI looks green now 👏

Indeed, it is 🎉!

@catenacyber has this been resolved:

> PS [@Yonah125](https://github.com/Yonah125) do you want to co-author commits with your comrades ?

Because I think we're ready to merge this PR once you're all happy with the commit history.

**Comment by catenacyber** [comment]
Commit history is good enough for me :-)

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**
